octobot_trading/personal_data/positions/position.py

```diff
@@ -281,22 +304,56 @@ def _is_update_increasing_size(self, size_update):
         if self.is_idle():
             return True
         if self.is_long():
-            return size_update > 0
-        return size_update < 0
+            return size_update > constants.ZERO
+        return size_update < constants.ZERO
+
+    def _is_update_closing(self, size_update):
+        """
+        :param size_update: the size update
+        :return: True if this update will close the position
+        """
+        if self.is_idle():
+            return True
+        if self.is_long():
+            return self.size + size_update <= constants.ZERO
+        return self.size + size_update >= constants.ZERO
 
     def _update_size(self, update_size):
         """
         Updates position size and triggers size related attributes update
         :param update_size: the size quantity
+        :return: True if the update increased position size
         """
+        is_update_increasing_position_size = self._is_update_increasing_size(update_size)
+        if not is_update_increasing_position_size:
+            self._update_realized_pnl_from_size_update(update_size, is_closing=self._is_update_closing(update_size))
         self._check_and_update_size(update_size)
         self._update_quantity()
         self._update_side()
-        self.update_initial_margin()
-        self.update_fee_to_close()
-        self.update_liquidation_price()
-        self.update_value()
-        self.update_pnl()
+        if self.exchange_manager.is_simulated:
+            self.update_initial_margin()
+            self.update_fee_to_close()
+            self.update_liquidation_price()
+            self.update_value()
+            self.update_pnl()
```

Comment:
these are already provided by the exchange when not in simulation ?

Reply:
Yes. It's useless to calculate twice pnl, liq price, etc...